pkg/querier/queryrange/index_stats_cache.go

```diff
@@ -0,0 +1,56 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+)
+
+type IndexStatsSplitter struct{}
+
+// GenerateCacheKey generates a cache key based on the userID, Request and interval.
+func (i IndexStatsSplitter) GenerateCacheKey(_ context.Context, userID string, r queryrangebase.Request) string {
+	return fmt.Sprintf("indexStats:%s:%s:%d:%d:%d", userID, r.GetQuery(), r.GetStep(), r.GetStart(), r.GetEnd())
+}
+
+type IndexStatsExtractor struct{}
+
+func (p IndexStatsExtractor) Extract(start, end int64, from queryrangebase.Response) queryrangebase.Response {
+	return from.(*IndexStatsResponse)
```

Comment:
Extract expects to be able to extract a subset of a response based on start & end. I'd suggest guessing what this should be by taking a proportion of the stats request based on the what % of the total request time the subset range covers.

Reply:
Done https://github.com/grafana/loki/pull/8972/commits/fffbc332702263b70eb5bbbbbbc2827744d81cc1